crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2538,29 +2571,51 @@ impl AuthorityPerEpochStore {
             Some(rm) => Some(rm.lock().await),
             None => None,
         };
-        let dkg_closed = self.randomness_state_enabled()
+        let dkg_failed = self.randomness_state_enabled()
             && randomness_manager
                 .as_ref()
                 .expect("randomness manager should exist if randomness is enabled")
-                .is_dkg_closed();
-        if dkg_closed {
-            let deferred_randomness_txs =
-                self.load_deferred_transactions_for_randomness(&mut batch)?;
-            previously_deferred_tx_digests.extend(deferred_randomness_txs.iter().flat_map(
-                |(deferral_key, txs)| {
-                    txs.iter().map(|tx| match tx.0.transaction.key() {
-                        SequencedConsensusTransactionKey::External(
-                            ConsensusTransactionKey::Certificate(digest),
-                        ) => (digest, *deferral_key),
-                        _ => panic!(
-                            "deferred randomness transaction was not a user certificate: {tx:?}"
-                        ),
-                    })
-                },
-            ));
-            sequenced_randomness_transactions
-                .extend(deferred_randomness_txs.into_iter().flat_map(|(_, txs)| txs));
+                .is_dkg_failed();
+        if dkg_failed {
+            self.load_and_process_deferred_transactions_for_randomness(
+                &mut batch,
+                &mut previously_deferred_tx_digests,
+                &mut sequenced_randomness_transactions,
+            )?;
         }
+
+        // Generate randomness for this commit if:
+        // 1. randomness is enabled
+        // 2. DKG is completed successfully
+        // 3. we are still accepting certs
+        let randomness_round = if !dkg_failed
+            && self.randomness_state_enabled()
+            && randomness_manager
+                .as_ref()
+                .expect("randomness manager should exist if randomness is enabled")
+                .is_dkg_successful()
+            // It is ok to just release lock here as functions called by this one are the
+            // only place that transition into RejectAllCerts state, and this function
+            // itself is always executed from consensus task.
+            && self
+                .get_reconfig_state_read_lock_guard()
+                .should_accept_tx()
+        {
+            randomness_manager
+                .as_mut()
+                .expect("randomness manager should exist if randomness is enabled")
```

Comment:
this check seems duplicated

Reply:
Which check? If you mean the is_dkg_successful, I think it's not duplicated because in case !dkg_failed, the DKG could either be successful or still in-progress.